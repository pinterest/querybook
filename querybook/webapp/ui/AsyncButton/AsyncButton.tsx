import {
    ButtonProps,
    ButtonType,
    getButtonComponentByType,
} from '../Button/Button';
import React from 'react';
import clsx from 'clsx';

export interface IAsyncButtonProps extends ButtonProps {
    onClick: (...args: any[]) => Promise<unknown>;
    disableWhileAsync?: boolean;
    type?: ButtonType;
}

export interface IAsyncButtonState {
    loading?: boolean;
}

export class AsyncButton extends React.PureComponent<
    IAsyncButtonProps,
    IAsyncButtonState
> {
    public static defaultProps: Partial<IAsyncButtonProps> = {
        disableWhileAsync: true,
    };
    private canSetState: boolean = true;

    public constructor(props: IAsyncButtonProps) {
        super(props);

        this.state = {
            loading: false,
        };
    }

    public componentWillUnmount() {
        this.canSetState = false;
    }

    public asyncSetState(newState: Partial<IAsyncButtonState>) {
        return new Promise<void>((resolve) => {
            if (this.canSetState) {
                this.setState(newState, resolve);
            }
        });
    }

    public onClick = async (...args: any[]) => {
        const { disableWhileAsync } = this.props;
        const { loading } = this.state;
        if (!(loading && disableWhileAsync)) {
            await this.asyncSetState({ loading: true });
            try {
                await this.props.onClick(...args);
            } finally {
                await this.asyncSetState({ loading: false });
            }
        }
    };

    public render() {
        const { disableWhileAsync, type, ...propsForButton } = this.props;
        const { loading } = this.state;

        const buttonProps: ButtonProps = {
            ...propsForButton,

            onClick: this.onClick,
            isLoading: loading,
            className: clsx({
                [propsForButton.className || '']: true,
            }),
        };
        const Button = getButtonComponentByType(type);
        return <Button {...buttonProps}>{this.props.children}</Button>;
    }
}
