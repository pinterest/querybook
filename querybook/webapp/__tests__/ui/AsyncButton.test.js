import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';

import { AsyncButton } from '../../ui/AsyncButton/AsyncButton';

it('renders without crashing', () => {
    shallow(<AsyncButton title="test" onClick={() => null} />);
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(
            <AsyncButton title="test" onClick={() => null} />
        );
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
    it('matches snapshot - disbled', () => {
        const wrapper = shallow(
            <AsyncButton title="test" onClick={() => null} disabled />
        );
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});

describe('matches test renderer snapshot', () => {
    it('serializes the styles', () => {
        const output = renderer.create(
            <AsyncButton title="test" onClick={() => null} />
        );
        expect(output).toMatchSnapshot();
    });
});
