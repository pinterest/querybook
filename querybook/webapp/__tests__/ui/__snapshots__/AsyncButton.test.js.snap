// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot - disbled 1`] = `
<Button
  className=""
  color="default"
  disabled={true}
  isLoading={false}
  onClick={[Function]}
  size="medium"
  theme="outline"
  title="test"
/>
`;

exports[`matches enzyme snapshots matches snapshot 1`] = `
<Button
  className=""
  color="default"
  isLoading={false}
  onClick={[Function]}
  size="medium"
  theme="outline"
  title="test"
/>
`;

exports[`matches test renderer snapshot serializes the styles 1`] = `
<span
  className="StyledButton-aqwbwz-0 eZYUp Button"
  color="var(--text-color)"
  onClick={[Function]}
  size="medium"
>
  <span>
    test
  </span>
</span>
`;
